### PR TITLE
Cairo AIR with constraints

### LINKED ISSUE #253: Cairo AIR with constraints
[Source](https://github.com/maxgillett/giza/blob/master/air/src/constraints.rs)

We need to check that the math done in this repo matches with the math done in the [paper](https://eprint.iacr.org/2021/1063.pdf).

### PR DESCRIPTION
**Pull request by gabrielbosio**
# TITLE

## Description

Resolves #253.

## Type of change

Please delete options that are not relevant.

- [ ] New feature

## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [ ] Benchmarks added/run

### CODE REVIEW

File: proving_system/stark/src/air/example/cairo.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        - &curr[FRAME_OP1_ADDR];
+}
+
+fn compute_register_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    let curr = frame.get_row(0);
+    let next = frame.get_row(1);
+
+    let one = FE::one();
+    let two = FE::from(2);
+
```
Comment (Oppen) [inline proving_system/stark/src/air/example/cairo.rs:248]: You should check the slice is big enough before starting indexing, otherwise it can panic.

File: math/src/helpers.rs [hunk 2]
```diff
@@ -1,7 +1,19 @@
...
         1
     } else {
         (u64::MAX >> (n - 1).leading_zeros()) + 1
     }
 }
+
+pub fn resize_to_next_power_of_two<F: IsTwoAdicField>(trace_colums: &mut [Vec<FieldElement<F>>]) {
+    trace_colums.iter_mut().for_each(|col| {
+        // TODO: Remove this unwrap. This may panic if the usize cant be
+        // casted into a u64.
```
Comment (Oppen) [inline math/src/helpers.rs:14]: You have at least a few years of runway before this is a problem. `usize` to `u64` conversion can only fail if the machine word exceeds 64 bits. I don't know of any architectures for which that is true yet.

**Review by MauroToscano — APPROVED** [review]

File: proving_system/stark/src/air/example/cairo.rs [hunk 3]
```diff
@@ -0,0 +1,305 @@
+use lambdaworks_math::field::{
```
Comment (schouhy) [inline proving_system/stark/src/air/example/cairo.rs:1]: I would name this module `cairo` instead of `example`

File: proving_system/stark/src/air/example/cairo.rs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+fn enforce_selector(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    let curr = frame.get_row(0);
+    for result_cell in constraints.iter_mut().take(ASSERT_EQ + 1).skip(INST) {
+        *result_cell = result_cell.clone() * curr[FRAME_SELECTOR].clone();
+    }
+}
+
+fn frame_inst_size(frame_row: &[FE]) -> FE {
+    &frame_row[F_OP_1_VAL] + FE::one()
+}
```
Comment (schouhy) [inline proving_system/stark/src/air/example/cairo.rs:305]: We could add here a test that imports the trace of a simple program and checks that all constraints are zero.

File: proving_system/stark/src/air/example/cairo.rs [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        };
+    }
+
+    // Instruction unpacking
+    let two = FE::from(2);
+    let b16 = two.pow(16u32);
+    let b32 = two.pow(32u32);
+    let b48 = two.pow(48u32);
+
+    let f0 = &curr[0..15]
```
Comment (schouhy) [inline proving_system/stark/src/air/example/cairo.rs:205]: Can we call this `f0_tilde` or `f0_squiggle` to match the notation of the paper?

**Review by schouhy — APPROVED** [review]
Awesome! I left a few minor comments. Maybe the comment about the test would be nice to have.

File: proving_system/stark/src/air/example/cairo.rs (hunk 5 above)
Comment (entropidelic) [reply proving_system/stark/src/air/example/cairo.rs:205]: good suggestion, done!

File: proving_system/stark/src/air/example/cairo.rs (hunk 4 above)
Comment (entropidelic) [reply proving_system/stark/src/air/example/cairo.rs:306]: done, thanks!

File: proving_system/stark/src/air/example/cairo.rs (hunk 3 above)
Comment (entropidelic) [reply proving_system/stark/src/air/example/cairo.rs:1]: I agree, lets merge this now and do that later 👍
